core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -56,263 +54,228 @@
+    if (listenerToClose != null) {
+      listenerToClose.closed(reason, new Metadata());
     }
   }
 
   @Override
   public void halfClose() {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingHalfClose = true;
-          return;
-        }
+    delayOrExecute(new Runnable() {
+      @Override
+      public void run() {
+        realStream.halfClose();
       }
-    }
-    startedRealStream.halfClose();
+    });
   }
 
   @Override
-  public void request(int numMessages) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingFlowControlRequests += numMessages;
-          return;
+  public void request(final int numMessages) {
+    if (passThrough) {
+      realStream.request(numMessages);
+    } else {
+      delayOrExecute(new Runnable() {
+        @Override
+        public void run() {
+          realStream.request(numMessages);
         }
-      }
+      });
     }
-    startedRealStream.request(numMessages);
   }
 
   @Override
-  public void setCompressor(Compressor compressor) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          this.compressor = compressor;
-          return;
-        }
+  public void setCompressor(final Compressor compressor) {
+    checkNotNull(compressor, "compressor");
+    delayOrExecute(new Runnable() {
+      @Override
+      public void run() {
+        realStream.setCompressor(compressor);
       }
-    }
-    startedRealStream.setCompressor(compressor);
+    });
   }
 
   @Override
   public void setDecompressor(Decompressor decompressor) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          this.decompressor = decompressor;
-          return;
-        }
-      }
-    }
-    startedRealStream.setDecompressor(decompressor);
+    checkNotNull(decompressor, "decompressor");
+    // This method being called only makes sense after setStream() has been called (but not
+    // necessarily returned), but there is not necessarily a happens-before relationship. This
+    // synchronized block creates one.
+    synchronized (this) { }
+    checkState(realStream != null, "How did we receive a reply before the request is sent?");
+    // ClientStreamListenerImpl (in ClientCallImpl) requires setDecompressor to be set immediately,
+    // since messages may be processed immediately after this method returns.
+    realStream.setDecompressor(decompressor);
   }
 
   @Override
   public boolean isReady() {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          return false;
-        }
-      }
+    if (passThrough) {
+      return realStream.isReady();
+    } else {
+      return false;
     }
-    return startedRealStream.isReady();
   }
 
   @Override
```

Comment:
I don't understand what you want to be pointed out. You want me to put a comment saying "not thread safe"?